client/cli-derive/examples/basic.rs

```diff
@@ -0,0 +1,25 @@
+#![allow(unused_variables, dead_code)]
+
+use sc_cli::{CliConfiguration, ImportParams, KeystoreParams, SharedParams};
+use sc_cli_derive::{spec_factory, substrate_cli_params};
+
+struct MyCli {
+	shared: SharedParams,
+	import: ImportParams,
+	keystore: KeystoreParams,
+}
+
+#[spec_factory(
+	cli = MyCli,
+	support_url = "http://example.org",
+	copyright_start_year = 2020,
+	impl_version = "0.1.0",
+)]
+fn spec_factory(id: &str) -> Result<Box<dyn sc_service::ChainSpec>, String> {
+	Err("not implemented".into())
+}
+
+#[substrate_cli_params(shared_params = shared, import_params = import, keystore_params = keystore)]
```

Comment:
What screams to be when looking at this example is that the procedural macro should really be a custom derive. They are the idiomatic way to automatically implement a trait on a struct.
An API like this looks much more comprehensible to me:

```rust
#[derive(CliCommand)]
#[cli_command(spec_factory = spec_factory, support_url = "foo", ...)]
struct MyCli {
        #[cli_command(shared_params)]
	shared: SharedParams,
        #[cli_command(import_params)]
	import: ImportParams,
        #[cli_command(keystore_params)]
	keystore: KeystoreParams,
}

// No attribute here
fn spec_factory(id: &str) -> Result<Box<dyn sc_service::ChainSpec>, String> {
	Err("not implemented".into())
}

// No manual `impl`
```

Reply:
TL;DR: I will work more on this to see what I can do.

Long:

I actually tried this pattern but there are multiple problems with it:

1. it makes the cli.rs not agnostic of substrate. We currently use the cli.rs to build auto-completion following this pattern: https://clap.rs/2016/10/25/an-update/ I splitted cli.rs into cli.rs + command.rs explicitly for that reason. See also: https://github.com/paritytech/substrate/blob/26e37ff/bin/node/cli/build.rs

    (Note that I don't recall what exactly was the problem if we don't split... I think it was a compilation problem with WASM)

2. substrate_cli_params actually allows to complete and override any field. Example [here](https://github.com/paritytech/substrate/blob/dc7703c/utils/frame/benchmarking-cli/src/command.rs#L128-L135) where the chain_spec is loaded differently for that particular subcommand.

3. having spec_factory on the root structopt is actually easier so I don't have to pass the function around. See how [Runtime pass it around](https://github.com/paritytech/substrate/blob/dc7703c6bd653d8abf75bdd6046df7511c54e005/client/cli/src/runtime.rs#L121) to create the Configuration.

Reply:
I couldn't find a better idea for now. I do like the derive macro but I know for sure it doesn't work (tested that first) because mostly of the point 2. I will close it for now but I'm opened to suggestion.

Meanwhile I did change spec_factory macro:
1. It now must be used on an impl (exactly like substrate_cli_params, this is better for consistency), it's also more explicit
2. I renamed it to substrate_cli_configuration since what it does is to implement substrate_cli_configuration for you